app/src/main/java/com/example/sharingang/MessageFragment.kt

```diff
@@ -0,0 +1,226 @@
+        messagesLiveData.observe(viewLifecycleOwner, { newList ->
+            (binding.history.adapter as MessageAdapter).submitList(newList)
+        })
+        listChats = mutableListOf()
+        partnerId = args.partnerId
+        partnerUsername = args.partnerUsername
+        partnerProfilePic = args.partnerProfilePictureUrl
+        currentUserId = currentUserProvider.getCurrentUserId()!!
+        setupFields()
+        setupSendButton()
+        setupUI()
+        setupMessageRefresh()
+        return binding.root
+    }
+
+    /**
+     * Setup the views for the profile picture and the username
+     * of the chat partner on the top of the fragment.
+     */
+    private fun setupFields() {
+        Glide.with(this).load(partnerProfilePic).into(binding.chatPartnerPic)
+        binding.chatPartnerUsername.text = partnerUsername
+    }
+
+    /**
+     * Setup the View and action for the button used for sending a message.
+     */
+    private fun setupSendButton() {
+        binding.btnSend.isEnabled = false
+        binding.messageEditText.addTextChangedListener {
+            binding.btnSend.isEnabled = !it.isNullOrBlank()
+        }
+        binding.btnSend.setOnClickListener {
+            val message: String = binding.messageEditText.text.toString()
+            sendMessage(currentUserId, partnerId, message)
+        }
+    }
+
+    /**
+     * Retrieve the user's document from the database based on its id.
+     *
+     * @param userId the id of the user we want to get the document of
+     * @return the document of the user
+     */
+    private fun getUserDocument(userId: String): DocumentReference {
+        val usersCollection = firebaseFirestore.collection(getString(R.string.users))
+        return usersCollection.document(userId)
+    }
+
+    /**
+     * Puts a message into the database and clears the message text field.
+     *
+     * @param from the sender
+     * @param to the receiver
+     * @param message the message to send
+     */
+    private fun sendMessage(from: String, to: String, message: String) {
+        listChats.clear()
+        val data = hashMapOf<String, Any>(
+            getString(R.string.message) to message,
+            getString(R.string.from) to from,
+            getString(R.string.to) to to
+        )
+        val lastTimeChat = hashMapOf(
+            getString(R.string.last_message) to "${Date()} (${System.currentTimeMillis()})"
+        )
+        val fromToPair = LinkedPair(from, to)
+        listOf(from, to).forEach {
+            val userDocument = getUserDocument(it)
+            userDocument.collection(getString(R.string.chats)).document(fromToPair.otherOf(it)!!)
+                .collection(getString(R.string.messages)).document(Date().toString()).set(data)
+            userDocument.collection(getString(R.string.messagePartners))
+                .document(fromToPair.otherOf(it)!!).set(lastTimeChat)
+        }
```

Comment:
You could save the current date in a variable and then reuse it instead of doing Date(). This way you are sure that the date is the same and not off by a few milliseconds.